cabal-install/Distribution/Client/ProjectOrchestration.hs

```diff
@@ -223,8 +223,13 @@ data ProjectBuildContext = ProjectBuildContext {
 runProjectPreBuildPhase
     :: Verbosity
     -> ProjectBaseContext
-    -> (ElaboratedInstallPlan -> IO ElaboratedInstallPlan)
-    -> IO ProjectBuildContext
+    -- ^ A function that selects a subset of the plan and returns it,
+    -- It can also return some arbitrary data, which can be useful in case
+    -- the caller needs it.
+    -> (ElaboratedInstallPlan -> IO (ElaboratedInstallPlan, a))
```

Comment:
I'm wary of over-abstracting the code prematurely for the sake of future consumers that may never materialise. If they do, refactoring is easy in Haskell. Existing type is nice because there's only one place you need to look at to see the data that's produced in the pre-build step.

> But then I'd have to unwrap that `Maybe` and rely on the assumption that I set it to  `Just` in preBuild.

That's okay IMO, just make it a fatal error if you get `Nothing`.